File: framework/python/betterportal/media.py
"""BP representation selection over parsed HTTP media ranges."""
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Iterable

from .contracts import contract

_MIME = {"json": "application/json", "html": "text/html", "metadata": "application/vnd.betterportal.metadata+json", "ndjson": "application/x-ndjson"}
_QUALITY = re.compile(r"(?:0(?:\.[0-9]{0,3})?|1(?:\.0{0,3})?)\Z")
_TOKEN = r"[!#$%&'*+.^_`|~A-Za-z0-9-]+"
_PARAM = re.compile(rf';[ \t]*({_TOKEN})[ \t]*=[ \t]*({_TOKEN}|"(?:[\t !#-\[\]-~\x80-\xff]|\\[\t !-~\x80-\xff])*")[ \t]*')
_RANGE = re.compile(rf"({_TOKEN}/{_TOKEN})[ \t]*")


def _ranges(raw: str):
    # email's MIME parser repairs malformed parameters and is not an HTTP parser.
    position = 0
    while position < len(raw):
        if raw[position] in " \t,":
            position += 1
            continue
        match = _RANGE.match(raw, position)
        if match is None:
            raise NotAcceptable("Invalid media range")
        media, position = match[1].lower(), match.end()
        if media.startswith("*/") and media != "*/*":
            raise NotAcceptable("Invalid media wildcard")
        params = {}
        while position < len(raw) and raw[position] == ";":
            parameter = _PARAM.match(raw, position)
            if parameter is None or parameter[1].lower() in params:
                raise NotAcceptable("Invalid media parameters")
            params[parameter[1].lower()] = parameter[2]
            position = parameter.end()
        if position < len(raw) and raw[position] != ",":
            raise NotAcceptable("Invalid media range")
        yield media, params


class NotAcceptable(ValueError):
    status = 406


@dataclass(frozen=True)
class Representation:
    kind: str
    mode: str | None = None


def negotiate(accept: str | None, available: Iterable[str] = tuple(_MIME)) -> Representation:
    """Honor q=0 and more-specific exclusions, then quality and request order."""
    supported = set(available)
    if not supported <= _MIME.keys():
        raise ValueError("Unknown representation")
    raw = accept if accept and accept.strip() else "*/*"
    if len(raw) > 8192 or any(ord(char) < 32 and char != "\t" or ord(char) == 127 for char in raw):
        raise NotAcceptable("Invalid Accept header")
    entries = []
    for index, (media, params) in enumerate(_ranges(raw)):
        quality_text = params.pop("q", "1")
        if not _QUALITY.fullmatch(quality_text):
            raise NotAcceptable("Invalid media quality")
        mode = params.get("mode")
        if mode is not None and mode.startswith('"'):
            mode = re.sub(r"\\(.)", r"\1", mode[1:-1])
        if mode is not None and not contract("RenderModeSchema").safe_parse(mode).success:
            mode = "invalid"
        entries.append((media, float(quality_text), mode, index))
    candidates = []
    for kind, mime in _MIME.items():
        if kind not in supported:
            continue
        for mode in ("page", "fragment", "embed") if kind == "html" else (None,):
            matches = []
            for media, quality, requested_mode, index in entries:
                if kind == "html" and (requested_mode or "page") != mode:
                    continue
                specificity = 2 if media == mime else 1 if media == mime.split("/")[0] + "/*" else 0 if media == "*/*" else -1
                if specificity >= 0:
                    matches.append((specificity + (1 if kind == "html" and requested_mode is not None else 0), quality, -index))
            if matches:
                _, quality, order = max(matches)
                if quality > 0:
                    candidates.append((quality, order, Representation(kind, mode)))
    if not candidates:
        raise NotAcceptable("No acceptable representation")
    return max(candidates, key=lambda candidate: candidate[:2])[2]

File: framework/dotnet/BetterPortal/Media.cs
using System.Text.RegularExpressions;

namespace BetterPortal;

public sealed class NotAcceptableException(string message) : Exception(message)
{
    public int Status => 406;
}
public sealed record Representation(string Kind, string? Mode = null);

public static class Media
{
    private static readonly Dictionary<string, string> Mime = new()
    {
        ["json"] = "application/json", ["html"] = "text/html",
        ["metadata"] = "application/vnd.betterportal.metadata+json", ["ndjson"] = "application/x-ndjson"
    };
    public static Representation Negotiate(string? accept, IEnumerable<string>? available = null)
    {
        var supported = (available ?? Mime.Keys).ToHashSet(StringComparer.Ordinal);
        if (!supported.IsSubsetOf(Mime.Keys)) throw new ArgumentException("Unknown representation");
        var raw = string.IsNullOrWhiteSpace(accept) ? "*/*" : accept;
        if (raw.Length > 8192 || raw.Any(character => character < 32 && character != '\t' || character == 127))
            throw new NotAcceptableException("Invalid Accept header");
        // HttpHeaders tolerates whitespace around '/', which is not media-type grammar.
        if (Regex.Matches(raw, "\"(?:[^\"\\\\]|\\\\.)*\"|(?<bad>[ \\t]/|/[ \\t])").Any(match => match.Groups["bad"].Success))
            throw new NotAcceptableException("Invalid media range");
        using var request = new HttpRequestMessage();
        try { request.Headers.Accept.ParseAdd(raw); }
        catch (FormatException) { throw new NotAcceptableException("Invalid media range"); }
        var entries = request.Headers.Accept.Select((entry, index) =>
        {
            if (entry.MediaType!.StartsWith("*/", StringComparison.Ordinal) && entry.MediaType != "*/*")
                throw new NotAcceptableException("Invalid media wildcard");
            var parameters = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            foreach (var parameter in entry.Parameters)
                if (parameter.Value is null || !parameters.TryAdd(parameter.Name, parameter.Value)) throw new NotAcceptableException("Invalid media parameters");
            var quality = parameters.GetValueOrDefault("q", "1");
            if (!Regex.IsMatch(quality, @"\A(?:0(?:\.[0-9]{0,3})?|1(?:\.0{0,3})?)\z")) throw new NotAcceptableException("Invalid media quality");
            var mode = parameters.GetValueOrDefault("mode");
            if (mode is { Length: >= 2 } && mode[0] == '"' && mode[^1] == '"') mode = Regex.Replace(mode[1..^1], @"\\(.)", "$1");
            if (mode is not null && !Contracts.Get("RenderModeSchema").SafeParse(mode).Success) mode = "invalid";
            return (Media: entry.MediaType!.ToLowerInvariant(), Quality: double.Parse(quality, System.Globalization.CultureInfo.InvariantCulture), Mode: mode, Index: index);
        }).ToArray();
        var candidates = new List<(double Quality, int Order, Representation Value)>();
        foreach (var (kind, mime) in Mime)
        {
            if (!supported.Contains(kind)) continue;
            foreach (var mode in kind == "html" ? new[] { "page", "fragment", "embed" } : new string?[] { null })
            {
                var matches = entries.Where(entry => kind != "html" || (entry.Mode ?? "page") == mode)
                    .Select(entry => (entry, specificity: entry.Media == mime ? 2 : entry.Media == mime.Split('/')[0] + "/*" ? 1 : entry.Media == "*/*" ? 0 : -1))
                    .Where(match => match.specificity >= 0)
                    .OrderByDescending(match => match.specificity + (kind == "html" && match.entry.Mode is not null ? 1 : 0))
                    .ThenByDescending(match => match.entry.Quality).ThenBy(match => match.entry.Index).ToArray();
                if (matches.Length > 0 && matches[0].entry.Quality > 0)
                    candidates.Add((matches[0].entry.Quality, matches[0].entry.Index, new(kind, mode)));
            }
        }
        return candidates.OrderByDescending(candidate => candidate.Quality).ThenBy(candidate => candidate.Order).FirstOrDefault().Value
            ?? throw new NotAcceptableException("No acceptable representation");
    }
}
